transport/udp/udp.go

```diff
@@ -0,0 +1,390 @@
+// Copyright 2016 The Mangos Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use file except in compliance with the License.
+// You may obtain a copy of the license at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package udp implements the UDP transport for mangos. It is based on
+// M. Sustrik, Ed., "UDP Mapping for Scalability Protocols",
+// sp-udp-mapping-01 (March 2014).
+package udp
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/go-mangos/mangos"
+)
+
+// Errors produced by this transport prototocol
+var (
+	ErrMsgTruncated = errors.New("message truncated due to insufficient buffer space")
+)
+
+// connHeader is exchanged during the initial handshake.
+// FIXME: stolen from conn.go -- unify this
+type connHeader struct {
+	Zero    byte // must be zero
+	S       byte // 'S'
+	P       byte // 'P'
+	Version byte // only zero at present
+	Proto   uint16
+	Rsvd    uint16 // always zero at present
+}
+
+const (
+	// Maximum read queue depth for listening endpoints
+	// FIXME: instead of udpMaxBacklog, use a ListenerOption to set this
+	// FIXME: e.g. OptionUdpListenerBacklog
+	udpMaxBacklog = 100
+
+	// Maximum expected UDP datagram size.
+	// UDPv4 datagrams can be up to 65507 = 2^16-1 - 20 - 8 bytes, due to the 2-byte
+	// IPv4 total length field, a minimum IPv4 header size of 20 bytes, and the 8
+	// byte UDP header length. IPv6 minimum header size is 40 bytes, so it is slightly
+	// less (not considering UDPv6 jumbograms, RFC 2675).
+	// These are theoretical limits, since in practice the maximum datagram size
+	// is determined by the MTU (typically 1500 .. ~9000 bytes).
+	// Hence a buffer size of 64k should suffice for 'regular' networks.
+	udpMaxBuf = 1 << 16
```

Comment:
Have you had time to ponder this, and how would you like to proceed?

I am understanding your comment to mean that there may be a different way of integrating UDP support. The patch submitted for review presents one possible alternative I had explored until it worked reliably (in the REQ/REP example).

I am not claiming that this is the best possible solution. Rather I think that there is a deeper question: since the interface is already built around the `Listen`, `Dial` and `Accept` verbs, any integration of UDP will have relative pros and cons. 

There were good engineering reasons in the Berkely sockets API to use different verbs for record-/packet-oriented and for byte-stream oriented sockets; which still seem to hold today.
An attempt to cross the boundaries to me seems a bit like putting a triangular or hexagonal peg into a round hole.

In conclusion, I believe it might be better to approach the integration from a higher level, considering a packet-oriented interface for go-mangos.

But then I ask myself, would the change be worth it? Perhaps keep the interface as it is and keep it with the transports it already provides. One big advantage of that is that it is consistent.

Your thoughts?